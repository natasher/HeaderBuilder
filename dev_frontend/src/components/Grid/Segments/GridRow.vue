<template>
  <section :class="classObject">

    <header class="hbr-header" >
      <span class="hbr-title">{{ name }}</span>

      <a href = "#"
        class = "icon-edit"
        @click.passive = "$store.dispatch( 'modals/' + editRowHandler, { $modal: $modal })"
        v-if = "getGridStatus === 'custom'">
          <font-icon icon = "pencil" />
      </a>
    </header>

    <div class = "hbr-container">

      <!-- @slot Insert left cell, (also used for `all` cell) -->
      <slot name = "left" />
      <!-- @slot Insert left cell footer -->
      <slot name = "footer-left" />

      <!-- @slot Insert center cell -->
      <slot name = "center" />
      <!-- @slot Insert center footer -->
      <slot name = "footer-center"/>

      <!-- @slot Insert right cell -->
      <slot name = "right" />
      <!-- @slot Insert right footer -->
      <slot name = "footer-right"/>

    </div>

  </section>
</template>

<script>
import { mapGetters } from 'vuex'
import FontIcon from '../../BaseGeneric/FontIcon.vue'

export default {

  name: 'GridRow',

  components: {
    FontIcon,
  },

  props: {
    /**
     * Name of the row. Displayed as label on row hover.
     */
    name: {
      type    : String,
      required: true,
    },
    /**
     * Function handler determinates behaviour after
     * edit ( pencil ) icon was clicked
     */
    editRowHandler: {
      type    : String,
      required: true,
    },
  },

  computed: {
    ...mapGetters( 'ui', [ 'getGridStatus' ]),

    classObject: function () {
      return {
        'hb-row'    : true,
        'hbr-action': this.name === 'Action Bar',
        'hbr-first' : this.name === 'First Row',
        'hbr-second': this.name === 'Second Row',
      }
    },
  },

}
</script>