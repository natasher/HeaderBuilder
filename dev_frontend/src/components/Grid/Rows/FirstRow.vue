<template>
  <grid-row
    name           = "First Row"
    editRowHandler = "showFirstRowOptionsModal">

    <grid-row-cell
      slot    = "left"
      cellPos = "left"
      key     = "firstRowLeft"
      ref     = "firstRowLeft"
      row     = "firstRow" />

      <cell-footer
        slot = "footer-left"
        pos  = "left"
        key  = "left-middle-pos" >
          {{ leftOrRight ? 'align: middle' : 'align: left' }}
      </cell-footer>

    <grid-row-cell
      slot    = "center"
      cellPos = "center"
      key     = "firstRowCenter"
      ref     = "firstRowCenter"
      row     = "firstRow" />

      <cell-footer
        slot = "footer-center"
        pos  = "center"
        v-if = "topOrBottom" >
          align: center
      </cell-footer>

    <grid-row-cell
      slot    = "right"
      cellPos = "right"
      key     = "firstRowRight"
      ref     = "firstRowright"
      row     = "firstRow" />

      <cell-footer
        slot = "footer-right"
        pos  = "right"
        v-if = "topOrBottom" >
          align: right
      </cell-footer>


  </grid-row>
</template>

<script>
import { mapActions, mapGetters } from 'vuex'
import GridRow     from '../Segments/GridRow.vue'
import GridRowCell from '../Segments/GridRowCell.vue'
import CellFooter  from '../Segments/CellFooter.vue'

export default {

  name: 'FirstRow',

  components: {
    GridRow,
    GridRowCell,
    CellFooter,
  },

  computed: {
    ...mapGetters( 'ui', [ 'getLayoutPosition' ]),

    leftOrRight: function() {
      return (
        this.getLayoutPosition === 'left' || this.getLayoutPosition === 'right'
      )
    },

    topOrBottom: function() {
      return !this.leftOrRight
    },
  },

  methods: {
    ...mapActions( 'modals', [ 'showFirstRowOptionsModal' ]),
  },

}
</script>
