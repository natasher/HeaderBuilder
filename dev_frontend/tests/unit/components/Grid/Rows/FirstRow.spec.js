import { shallowMount, createLocalVue } from '@vue/test-utils'
import flushPromises from 'flush-promises'
import Vuex          from 'vuex'
import merge         from 'lodash.merge'
import FirstRow      from '../../../../../src/components/Grid/Rows/FirstRow.vue'
import GridRow       from '../../../../../src/components/Grid/Segments/GridRow.vue'

const localVue = createLocalVue()
localVue.use( Vuex )

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      ui: {
        namespaced: true,
        getters: {
          getLayoutPosition: jest.fn().mockReturnValue( 'top' ),
          getGridStatus    : jest.fn(() => { return 'custom' })
        }
      }
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    store: createStore(),
    stubs: {
      GridRow,
    }
  }

  return shallowMount( FirstRow, merge( defaultMountingOptions, overrides ))
}

describe('FirstRow.vue', () => {

  describe('snapshots:', () => {

    test('should match if the layout position is top', () => {
      const wrapper = createWrapper()

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match if the layout position is bottom', () => {
      const store = createStore({
        modules: {
          ui: {
            getters: {
              getLayoutPosition: jest.fn(() => { return 'bottom' })
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match if the layout position is right', () => {
      const store = createStore({
        modules: {
          ui: {
            getters: {
              getLayoutPosition: jest.fn(() => { return 'right' })
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match if the layout position is left', () => {
      const store = createStore({
        modules: {
          ui: {
            getters: {
              getLayoutPosition: jest.fn(() => { return 'left' })
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match if grid status is not custom', () => {
      const store = createStore({
        modules: {
          ui: {
            getters: {
              getGridStatus: jest.fn(() => { return 'auto' })
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })

  })

  test('should open actions bar options modal if edit icon is clicked', async () => {
    const $modal = jest.fn()
    const showFirstRowOptionsModal = jest.fn()
    const store = createStore({
      modules: {
        modals: {
          namespaced: true,
          actions: {
            showFirstRowOptionsModal,
          }
        }
      }
    })
    const wrapper = createWrapper({
      store,
      mocks: {
        $modal
      }
    })
    const editBtn = wrapper.find('.icon-edit')

    editBtn.trigger('click')
    await flushPromises()

    expect(showFirstRowOptionsModal).toBeCalledWith(
      expect.anything(),
      { $modal, },
      void 0
    )
  })

})