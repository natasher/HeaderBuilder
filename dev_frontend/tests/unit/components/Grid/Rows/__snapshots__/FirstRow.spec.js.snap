// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirstRow.vue snapshots: should match if grid status is not custom 1`] = `
<section
  class="hb-row hbr-first"
>
  <header
    class="hbr-header"
  >
    <span
      class="hbr-title"
    >
      First Row
    </span>
     
    <!---->
  </header>
   
  <div
    class="hbr-container"
  >
    <grid-row-cell-stub
      cellpos="left"
      row="firstRow"
    />
     
    <cell-footer-stub
      pos="left"
    >
      
        align: left
    
    </cell-footer-stub>
     
    <grid-row-cell-stub
      cellpos="center"
      row="firstRow"
    />
     
    <cell-footer-stub
      pos="center"
    >
      
        align: center
    
    </cell-footer-stub>
     
    <grid-row-cell-stub
      cellpos="right"
      row="firstRow"
    />
     
    <cell-footer-stub
      pos="right"
    >
      
        align: right
    
    </cell-footer-stub>
  </div>
</section>
`;

exports[`FirstRow.vue snapshots: should match if the layout position is bottom 1`] = `
<section
  class="hb-row hbr-first"
>
  <header
    class="hbr-header"
  >
    <span
      class="hbr-title"
    >
      First Row
    </span>
     
    <a
      class="icon-edit"
      href="#"
    >
      <font-icon-stub
        icon="pencil"
      />
    </a>
  </header>
   
  <div
    class="hbr-container"
  >
    <grid-row-cell-stub
      cellpos="left"
      row="firstRow"
    />
     
    <cell-footer-stub
      pos="left"
    >
      
        align: left
    
    </cell-footer-stub>
     
    <grid-row-cell-stub
      cellpos="center"
      row="firstRow"
    />
     
    <cell-footer-stub
      pos="center"
    >
      
        align: center
    
    </cell-footer-stub>
     
    <grid-row-cell-stub
      cellpos="right"
      row="firstRow"
    />
     
    <cell-footer-stub
      pos="right"
    >
      
        align: right
    
    </cell-footer-stub>
  </div>
</section>
`;

exports[`FirstRow.vue snapshots: should match if the layout position is left 1`] = `
<section
  class="hb-row hbr-first"
>
  <header
    class="hbr-header"
  >
    <span
      class="hbr-title"
    >
      First Row
    </span>
     
    <a
      class="icon-edit"
      href="#"
    >
      <font-icon-stub
        icon="pencil"
      />
    </a>
  </header>
   
  <div
    class="hbr-container"
  >
    <grid-row-cell-stub
      cellpos="left"
      row="firstRow"
    />
     
    <cell-footer-stub
      pos="left"
    >
      
        align: middle
    
    </cell-footer-stub>
     
    <grid-row-cell-stub
      cellpos="center"
      row="firstRow"
    />
      
    <grid-row-cell-stub
      cellpos="right"
      row="firstRow"
    />
     
  </div>
</section>
`;

exports[`FirstRow.vue snapshots: should match if the layout position is right 1`] = `
<section
  class="hb-row hbr-first"
>
  <header
    class="hbr-header"
  >
    <span
      class="hbr-title"
    >
      First Row
    </span>
     
    <a
      class="icon-edit"
      href="#"
    >
      <font-icon-stub
        icon="pencil"
      />
    </a>
  </header>
   
  <div
    class="hbr-container"
  >
    <grid-row-cell-stub
      cellpos="left"
      row="firstRow"
    />
     
    <cell-footer-stub
      pos="left"
    >
      
        align: middle
    
    </cell-footer-stub>
     
    <grid-row-cell-stub
      cellpos="center"
      row="firstRow"
    />
      
    <grid-row-cell-stub
      cellpos="right"
      row="firstRow"
    />
     
  </div>
</section>
`;

exports[`FirstRow.vue snapshots: should match if the layout position is top 1`] = `
<section
  class="hb-row hbr-first"
>
  <header
    class="hbr-header"
  >
    <span
      class="hbr-title"
    >
      First Row
    </span>
     
    <a
      class="icon-edit"
      href="#"
    >
      <font-icon-stub
        icon="pencil"
      />
    </a>
  </header>
   
  <div
    class="hbr-container"
  >
    <grid-row-cell-stub
      cellpos="left"
      row="firstRow"
    />
     
    <cell-footer-stub
      pos="left"
    >
      
        align: left
    
    </cell-footer-stub>
     
    <grid-row-cell-stub
      cellpos="center"
      row="firstRow"
    />
     
    <cell-footer-stub
      pos="center"
    >
      
        align: center
    
    </cell-footer-stub>
     
    <grid-row-cell-stub
      cellpos="right"
      row="firstRow"
    />
     
    <cell-footer-stub
      pos="right"
    >
      
        align: right
    
    </cell-footer-stub>
  </div>
</section>
`;
